File: agentguard-backend/backend/app/schemas/traces.py
"""Pydantic schemas for request tracing endpoints."""

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field


class TraceListItem(BaseModel):
    """Proxy request summary for trace listing."""

    model_config = ConfigDict(from_attributes=True, populate_by_name=True)

    id: UUID
    method: str
    path: str
    model: str | None
    status_code: int | None = Field(serialization_alias="statusCode")
    latency_ms: int | None = Field(serialization_alias="latencyMs")
    input_tokens: int | None = Field(serialization_alias="inputTokens")
    output_tokens: int | None = Field(serialization_alias="outputTokens")
    cost_usd: float | None = Field(serialization_alias="costUsd")
    incident_count: int = Field(default=0, serialization_alias="incidentCount")
    created_at: datetime = Field(serialization_alias="createdAt")


class TraceListResponse(BaseModel):
    """Paginated list of traces."""

    items: list[TraceListItem]
    total: int


class TraceIncident(BaseModel):
    """Incident summary within a trace."""

    model_config = ConfigDict(from_attributes=True, populate_by_name=True)

    id: UUID
    severity: str
    category: str
    title: str
    status: str
    action_taken: str | None = Field(serialization_alias="actionTaken")
    created_at: datetime = Field(serialization_alias="createdAt")


class TraceDetailResponse(BaseModel):
    """Full trace detail with request/response and linked incidents."""

    model_config = ConfigDict(populate_by_name=True)

    id: UUID
    method: str
    path: str
    model: str | None
    status_code: int | None = Field(serialization_alias="statusCode")
    latency_ms: int | None = Field(serialization_alias="latencyMs")
    input_tokens: int | None = Field(serialization_alias="inputTokens")
    output_tokens: int | None = Field(serialization_alias="outputTokens")
    cost_usd: float | None = Field(serialization_alias="costUsd")
    request_body: str | None = Field(serialization_alias="requestBody")
    response_body: str | None = Field(serialization_alias="responseBody")
    incidents: list[TraceIncident]
    created_at: datetime = Field(serialization_alias="createdAt")
    updated_at: datetime = Field(serialization_alias="updatedAt")

File: agentguard-frontend/src/lib/api/traces.ts
/**
 * Trace (proxy request) API client.
 */

import { apiFetch, buildQueryString } from './client';

export interface TraceListItem {
  id: string;
  method: string;
  path: string;
  model: string | null;
  statusCode: number | null;
  latencyMs: number | null;
  inputTokens: number | null;
  outputTokens: number | null;
  costUsd: number | null;
  incidentCount: number;
  createdAt: string;
}

export interface TraceIncident {
  id: string;
  severity: string;
  category: string;
  title: string;
  status: string;
  actionTaken: string | null;
  createdAt: string;
}

export interface TraceDetail extends TraceListItem {
  requestBody: string | null;
  responseBody: string | null;
  incidents: TraceIncident[];
  updatedAt: string;
}

export interface TraceListResponse {
  items: TraceListItem[];
  total: number;
}

export interface TraceFilters {
  model?: string;
  statusCode?: number;
  hasIncidents?: boolean;
  q?: string;
  skip?: number;
  limit?: number;
}

export async function listTraces(filters: TraceFilters = {}): Promise<TraceListResponse> {
  const qs = buildQueryString(filters as Record<string, string | number | boolean | undefined>);
  return apiFetch<TraceListResponse>(`/traces${qs}`);
}

export async function getTrace(traceId: string): Promise<TraceDetail> {
  return apiFetch<TraceDetail>(`/traces/${traceId}`);
}
